Firestore/core/src/firebase/firestore/nanopb/reader.cc

```diff
@@ -36,22 +36,23 @@ Tag Reader::ReadTag() {
 
   bool eof;
   if (!pb_decode_tag(&stream_, &tag.wire_type, &tag.field_number, &eof)) {
-    status_ = Status(FirestoreErrorCode::DataLoss, PB_GET_ERROR(&stream_));
+    Fail(PB_GET_ERROR(&stream_));
     return tag;
   }
 
   // nanopb code always returns a false status when setting eof.
   HARD_ASSERT(!eof, "nanopb set both ok status and eof to true");
 
+  last_tag_ = tag;
   return tag;
 }
 
-bool Reader::RequireWireType(pb_wire_type_t wire_type, Tag tag) {
+bool Reader::RequireWireType(pb_wire_type_t wire_type) {
   if (!status_.ok()) return false;
-  if (wire_type != tag.wire_type) {
-    set_status(Status(FirestoreErrorCode::DataLoss,
-                      "Input proto bytes cannot be parsed (mismatch between "
-                      "the wiretype and the field number (tag))"));
+  if (wire_type != last_tag_.wire_type) {
+    Fail(
+        "Input proto bytes cannot be parsed (mismatch between the wiretype and "
```

Comment:
I worry that if this error were to happen to a customer we'd have essentially no way to debug it.

In the current Objective-C implementation where all errors are fatal, this kind of error is tolerable because it is accompanied by a stack trace and our recursive descent parsing of the protos makes it obvious at least in which message we failed.

Now that the reader will continue after erroring this message will essentially be context free if we ever see it. As a consequence this message, while detailed, is effectively no better than "something went wrong".

We should supply as much context as we can: last_tag_'s field number, wire type values, maybe even the containing message type if we can supply it.

I'm happy to take this in a separate PR, but this caught my eye and I figured I should mention it.